### PR TITLE
Struct for zoom level

### PR DESCRIPTION
**Pull request by ybiletskyi**
Created a struct for `ZoomLevel`

## ✔️ PR Todo

- [x] Included links to related issues/PRs
fixes https://github.com/maplibre/maplibre-rs/issues/95

### CODE REVIEW

File: maplibre/src/coords.rs [hunk 1]
```diff
@@ -33,12 +33,12 @@ const fn create_zoom_bounds<const DIM: usize>() -> [u32; DIM] {
...
-pub struct Quadkey([u8; MAX_ZOOM]);
+pub struct Quadkey([ZoomLevel; MAX_ZOOM]);
 
 impl Quadkey {
-    pub fn new(quad_encoded: &[u8]) -> Self {
-        let mut key = [0u8; MAX_ZOOM];
-        key[0] = quad_encoded.len() as u8;
+    pub fn new(quad_encoded: &[ZoomLevel]) -> Self {
+        let mut key = [ZoomLevel::default(); MAX_ZOOM];
+        key[0] = ZoomLevel::new(quad_encoded.len() as u8);
```
Comment (maxammann) [inline maplibre/src/coords.rs:41]: We could use the `into()` trait here.

File: maplibre/src/coords.rs [hunk 2]
```diff
@@ -48,14 +48,65 @@ impl Quadkey {
 
 impl fmt::Debug for Quadkey {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let len = self.0[0] as usize;
+        let len = self.0[0].0 as usize;
```
Comment (maxammann) [inline maplibre/src/coords.rs:51]: Minor:

```suggestion
        let Quadkey(key) = self.0;
        let ZoomLevel(level) = key[0];
        let len = level as usize;
```

File: maplibre/src/coords.rs [hunk 3]
```diff
@@ -48,14 +48,65 @@ impl Quadkey {
...
     }
 }
 
+#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Copy, Clone, Debug)]
+pub struct ZoomLevel(u8);
+
+impl ZoomLevel {
+    pub fn new(zoom_level: u8) -> Self {
+        ZoomLevel(zoom_level)
+    }
```
Comment (maxammann) [inline maplibre/src/coords.rs:65]: Should probably be removed becausee we have the `From` trait.

File: maplibre/src/coords.rs [hunk 4]
```diff
@@ -48,14 +48,65 @@ impl Quadkey {
...
+}
+
+impl std::ops::Mul<i32> for ZoomLevel {
+    type Output = ZoomLevel;
+
+    fn mul(self, rhs: i32) -> Self::Output {
+        ZoomLevel(self.0 * rhs as u8)
+    }
+}
```
Comment (maxammann) [inline maplibre/src/coords.rs:96]: Is this really used? I think mulitplication does not make sense for zoomlevels

File: maplibre/src/coords.rs [hunk 5]
```diff
@@ -481,15 +538,15 @@ impl ViewRegion {
     pub fn iter(&self) -> impl Iterator<Item = WorldTileCoords> + '_ {
         (self.min_tile.x - self.padding..self.max_tile.x + 1 + self.padding).flat_map(move |x| {
             (self.min_tile.y - self.padding..self.max_tile.y + 1 + self.padding).map(move |y| {
-                let tile_coord: WorldTileCoords = (x, y, self.z as u8).into();
+                let tile_coord: WorldTileCoords = (x, y, self.z.0 as u8).into();
```
Comment (maxammann) [inline maplibre/src/coords.rs:541]: We should construct a WorldTileCoords directly from the ZoomLevel. We should no unpack if first and then create a new one.

Maybe implement the From `(u32, u32, ZoomLevel)` trait for WorldTileCoords

File: maplibre/src/coords.rs [hunk 6]
```diff
@@ -318,7 +375,7 @@ impl WorldTileCoords {
 
     /// Get the tile which is one zoom level lower and contains this one
     pub fn get_parent(&self) -> Option<WorldTileCoords> {
-        if self.z == 0 {
+        if self.z == ZoomLevel::default() {
```
Comment (maxammann) [inline maplibre/src/coords.rs:378]: Lets add a function to ZoomLevel: `is_root()` which checks if it is `0`

**Review by maxammann — CHANGES_REQUESTED** [review]
Sorry for the picky review :D

> 10 lines of code = 10 issues.
> 500 lines of code = "looks fine."

File: maplibre/src/coords.rs (hunk 6 above)
Comment (ybiletskyi) [reply maplibre/src/coords.rs]: Fixed.

File: maplibre/src/coords.rs [hunk 7]
```diff
@@ -48,14 +48,65 @@ impl Quadkey {
...
+    fn add(self, rhs: i32) -> Self::Output {
+        ZoomLevel(self.0 + rhs as u8)
+    }
+}
+
+impl std::ops::Sub<i32> for ZoomLevel {
+    type Output = ZoomLevel;
+
+    fn sub(self, rhs: i32) -> Self::Output {
+        ZoomLevel(self.0 - rhs as u8)
```
[the remark this answers is not in the document]
Comment (ybiletskyi) [reply maplibre/src/coords.rs]: Added a checked call for operator.

File: maplibre/src/coords.rs (hunk 3 above)
Comment (ybiletskyi) [reply maplibre/src/coords.rs]: Yep. You are right. Deleted.

File: maplibre/src/coords.rs (hunk 1 above)
Comment (ybiletskyi) [reply maplibre/src/coords.rs]: Fixed.

**Review by maxammann — APPROVED** [review]

**Comment by maxammann** [comment]
android build is failing because of github actions, lets just merge this and see what happens

**Pull request merged**